File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceEvent.kt
package com.arcinteractive.spaces.data.model

import java.text.DateFormat
import java.util.Calendar
import java.util.Date
import java.util.TimeZone

data class SpaceEvent(
    val id: String,
    val spaceId: String,
    val title: String,
    val description: String,
    val location: String,
    val startDate: Date,
    val endDate: Date,
    val allDay: Boolean,
    val timezone: String,
    val createdBy: String,
    val createdByName: String,
    val createdAt: Date?,
    val updatedAt: Date?,
    val deleted: Boolean
) {
    val dateText: String
        get() = if (allDay) allDayDateFormatter.format(startDate) else dateFormatter.format(startDate)

    val timeText: String
        get() = if (allDay) {
            "All Day"
        } else {
            "${timeFormatter.format(startDate)} – ${timeFormatter.format(endDate)}"
        }
}

data class EditableSpaceEvent(
    val title: String = "",
    val description: String = "",
    val location: String = "",
    val startDate: Date = Date(),
    val endDate: Date = Calendar.getInstance().apply { time = Date(); add(Calendar.HOUR_OF_DAY, 1) }.time,
    val allDay: Boolean = false
) {
    val trimmedTitle: String
        get() = title.trim()

    val canSave: Boolean
        get() = trimmedTitle.isNotEmpty() && !endDate.before(startDate)

    companion object {
        fun fromEvent(event: SpaceEvent): EditableSpaceEvent {
            return EditableSpaceEvent(
                title = event.title,
                description = event.description,
                location = event.location,
                startDate = event.startDate,
                endDate = event.endDate,
                allDay = event.allDay
            )
        }
    }
}

sealed interface EventEditorMode {
    data object Create : EventEditorMode
    data class Edit(val event: SpaceEvent) : EventEditorMode

    val title: String
        get() = when (this) {
            Create -> "Add Event"
            is Edit -> "Edit Event"
        }
}

private val dateFormatter: DateFormat = DateFormat.getDateInstance(DateFormat.MEDIUM)
private val allDayDateFormatter: DateFormat = DateFormat.getDateInstance(DateFormat.FULL)
private val timeFormatter: DateFormat = DateFormat.getTimeInstance(DateFormat.SHORT)

private fun configureFormatters() {
    val zone = TimeZone.getDefault()
    dateFormatter.timeZone = zone
    allDayDateFormatter.timeZone = zone
    timeFormatter.timeZone = zone
}

@Suppress("unused")
private val eventFormatterConfig = run { configureFormatters() }

File: iOS/Spaces V1.5/Models/SpaceEvent.swift
import Foundation

struct SpaceEvent: Identifiable, Hashable {
    let id: String
    let spaceID: String
    let title: String
    let description: String
    let location: String
    let startDate: Date
    let endDate: Date
    let allDay: Bool
    let timeZoneIdentifier: String
    let createdBy: String
    let createdByName: String
    let createdAt: Date?
    let updatedAt: Date?
    let deleted: Bool
}

extension SpaceEvent {
    var timeZone: TimeZone {
        TimeZone(identifier: timeZoneIdentifier) ?? .current
    }

    var dateText: String {
        if allDay {
            return Self.allDayDateFormatter.string(from: startDate)
        }
        return Self.dateFormatter.string(from: startDate)
    }

    var timeText: String {
        guard !allDay else { return "All Day" }
        let start = Self.timeFormatter.string(from: startDate)
        let end = Self.timeFormatter.string(from: endDate)
        return "\(start) – \(end)"
    }

    var sortDate: Date {
        startDate
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .none
        return formatter
    }()

    private static let allDayDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMM d"
        return formatter
    }()

    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .none
        formatter.timeStyle = .short
        return formatter
    }()
}

struct EditableSpaceEvent {
    var title: String
    var description: String
    var location: String
    var startDate: Date
    var endDate: Date
    var allDay: Bool

    init(
        title: String = "",
        description: String = "",
        location: String = "",
        startDate: Date = Date(),
        endDate: Date = Calendar.current.date(byAdding: .hour, value: 1, to: Date()) ?? Date(),
        allDay: Bool = false
    ) {
        self.title = title
        self.description = description
        self.location = location
        self.startDate = startDate
        self.endDate = endDate
        self.allDay = allDay
    }

    init(event: SpaceEvent) {
        self.title = event.title
        self.description = event.description
        self.location = event.location
        self.startDate = event.startDate
        self.endDate = event.endDate
        self.allDay = event.allDay
    }

    var trimmedTitle: String {
        title.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var canSave: Bool {
        !trimmedTitle.isEmpty && endDate >= startDate
    }
}

enum EventEditorMode {
    case create
    case edit(SpaceEvent)

    var navigationTitle: String {
        switch self {
        case .create:
            return "Add Event"
        case .edit:
            return "Edit Event"
        }
    }
}
